test/hpke_test.c

```diff
@@ -0,0 +1,1728 @@
+        plainlen = strlen((char *)plain);
+        /*
+         * Randomly try with/without info, aad, seq. Given mode and suite
+         * combos, and this being run even a few times, we'll exercise many
+         * code paths fairly quickly. We don't really care what the values
+         * are but it'll be easier to debug if they're known, so we set 'em.
+         */
+        if (COIN_IS_HEADS) {
+            aadp = aad;
+            memset(aad, 'a', aadlen);
+        } else {
+            aadlen = 0;
+        }
+        if (COIN_IS_HEADS) {
+            infop = info;
+            memset(info, 'i', infolen);
+        } else {
+            infolen = 0;
+        }
+        if (hpke_mode == OSSL_HPKE_MODE_PSK
+            || hpke_mode == OSSL_HPKE_MODE_PSKAUTH) {
+            pskp = lpsk;
+            memset(lpsk, 'P', psklen);
+            pskidp = lpskid;
+            memset(lpskid, 'I', psklen - 1);
+            lpskid[psklen - 1] = '\0';
+        } else {
+            psklen = 0;
+        }
+        for (kemind = 0; /* iterate over the kems, kdfs and aeads */
+             overallresult == 1 && kemind != OSSL_NELEM(hpke_kem_list);
+             kemind++) {
+            uint16_t kem_id = hpke_kem_list[kemind];
+            size_t authpublen = OSSL_HPKE_TSTSIZE;
+            unsigned char authpub[OSSL_HPKE_TSTSIZE];
+            unsigned char *authpubp = NULL;
+            EVP_PKEY *authpriv = NULL;
+
+            hpke_suite.kem_id = kem_id;
+            if ((hpke_mode == OSSL_HPKE_MODE_AUTH) ||
+                (hpke_mode == OSSL_HPKE_MODE_PSKAUTH)) {
+                if (TEST_true(OSSL_HPKE_keygen(hpke_suite, authpub, &authpublen,
+                                               &authpriv, NULL, 0,
+                                               testctx, NULL)) != 1) {
+                    overallresult = 0;
+                }
+                authpubp = authpub;
+            } else {
+                authpublen = 0;
+            }
+            for (kdfind = 0;
+                 overallresult == 1 && kdfind != OSSL_NELEM(hpke_kdf_list);
+                 kdfind++) {
+                uint16_t kdf_id = hpke_kdf_list[kdfind];
+
+                hpke_suite.kdf_id = kdf_id;
+                for (aeadind = 0;
+                     overallresult == 1 && aeadind != OSSL_NELEM(hpke_aead_list);
+                     aeadind++) {
+                    uint16_t aead_id = hpke_aead_list[aeadind];
+                    size_t publen = OSSL_HPKE_TSTSIZE;
+                    unsigned char pub[OSSL_HPKE_TSTSIZE];
+                    size_t senderpublen = OSSL_HPKE_TSTSIZE;
+                    unsigned char senderpub[OSSL_HPKE_TSTSIZE];
+                    size_t cipherlen = OSSL_HPKE_TSTSIZE;
+                    unsigned char cipher[OSSL_HPKE_TSTSIZE];
+                    size_t clearlen = OSSL_HPKE_TSTSIZE;
+                    unsigned char clear[OSSL_HPKE_TSTSIZE];
+
+                    hpke_suite.aead_id = aead_id;
```

Comment:
I guess the question you have to ask yourself when you write a test with this many levels is,
How do I debug this if it fails at some point? (Probably wont happen - but it would be a pain if it does).

You could also add a test note in the innermost loop.. (Add somestring names in there also)
TEST_note("Testing blah=%d blah1=%d blah3=%d blah4=%d", mind, kemind, kdf_id, aeadind);

Reply:
Added a ``TEST_note()`` call like that for failure case or if verbose requested (using e.g. "V=1" on command line).

I copied how verbosity is handled in ``test/drbgtest.c`` which uses ``getenv("V")`` so if that's out of date/fashion, let me know and I can change to whatever's current.

Reply:
Code now uses ``opt_next()`` etc same as other recents.